File: apps/dashboard/frontend/src/models/billing-meter.ts
export type Tier = {
  start: number;
  end: number; // -1 means unlimited
  unit_amount: number; // in cents
};

export type BillingMeter = {
  id: string;
  subscription_id: string;
  feature_id: string;
  plan_item_id: string;
  user_id: string;
  usage: number;
  pricing_model: "unit" | "tiered" | "block" | "volume";
  unit_amount: number; // in cents
  free_quota: number;
  tiers?: Tier[];
  last_reset_at: string | null;
  next_reset_at: string | null;
  created_at: string;
  updated_at: string;

  // Computed/display fields
  projected_cost: number; // in cents
  feature_name?: string;
  plan_item_title?: string;
};

File: libs/go/services/entitlement/billing_meter.go
package entitlement

import (
	"github.com/useportcall/portcall/libs/go/dbx"
	"github.com/useportcall/portcall/libs/go/dbx/models"
)

func upsertBillingMeter(
	db dbx.IORM, sub *models.Subscription, pf *models.PlanFeature,
	featureID uint, evt *models.MeterEvent,
) error {
	var meter models.BillingMeter
	err := db.FindFirst(&meter,
		"subscription_id = ? AND plan_item_id = ? AND feature_id = ?",
		sub.ID, pf.PlanItemID, featureID,
	)
	if dbx.IsRecordNotFoundError(err) {
		return createBillingMeter(db, sub, pf, featureID, evt)
	}
	if err != nil {
		return err
	}

	meter.Usage = clampUsage(meter.Usage, evt.Usage)
	return db.Save(&meter)
}

func createBillingMeter(
	db dbx.IORM, sub *models.Subscription, pf *models.PlanFeature,
	featureID uint, evt *models.MeterEvent,
) error {
	var planItem models.PlanItem
	if err := db.FindForID(pf.PlanItemID, &planItem); err != nil {
		return err
	}

	interval, count := planItem.Interval, planItem.IntervalCount
	if interval == "" || interval == "inherit" {
		interval = sub.BillingInterval
		count = sub.BillingIntervalCount
	}

	freeQuota := int64(0)
	if pf.Quota > 0 {
		freeQuota = pf.Quota
	}

	meter := models.BillingMeter{
		AppID:          sub.AppID,
		SubscriptionID: sub.ID,
		PlanItemID:     planItem.ID,
		FeatureID:      featureID,
		UserID:         evt.UserID,
		Usage:          clampUsage(0, evt.Usage),
		PricingModel:   planItem.PricingModel,
		UnitAmount:     planItem.UnitAmount,
		Tiers:          planItem.Tiers,
		FreeQuota:      freeQuota,
		Interval:       interval,
		IntervalCount:  count,
		LastResetAt:    &sub.LastResetAt,
		NextResetAt:    &sub.NextResetAt,
	}
	return db.Create(&meter)
}
